lib/Sema/TypeChecker.cpp

```diff
@@ -496,6 +496,95 @@ void swift::typeCheckExternalDefinitions(SourceFile &SF) {
+// and complain!
+static void diagnoseUnsupportedCaptures(TypeChecker &TC, SourceFile &SF,
+  unsigned StartElem) {
+  class CallWalker: public ASTWalker {
+    TypeChecker &TC;
+    static unsigned countInOutCaputeres(Expr *E) {
+      unsigned inoutCount = 0;
+      if (auto DRE = dyn_cast<DeclRefExpr>(E))
+        if (auto FD = dyn_cast<FuncDecl>(DRE->getDecl()))
+          for (auto C: FD->getCaptureInfo().getCaptures()) {
+            if (isa<InOutType>(C.getDecl()->getInterfaceType().getPointer()))
+              inoutCount++;
+          }
+      return inoutCount;
+    }
+  public:
+    CallWalker(TypeChecker &TC): TC(TC) {}
+    std::pair<bool, Stmt *> walkToStmtPre(Stmt *S) {
+      if (auto RS = dyn_cast<ReturnStmt>(S)) {
+        if (RS->hasResult() && countInOutCaputeres(RS->getResult()))
+          TC.diagnose(RS->getReturnLoc(),
+            diag::nested_function_escaping_inout_capture);
+      }
+      return { true, S };
+    }
+    std::pair<bool, Expr *> walkToExprPre(Expr *E) {
+      if (auto CE = dyn_cast<ApplyExpr>(E)) {
+        // If it's not a function, or the function doesn't have any parameter,
+        // skip it.
+        MutableArrayRef<ParamDecl *> params;
+        if (auto DRE = dyn_cast<DeclRefExpr>(CE->getFn()))
+          if (auto FD = dyn_cast<FuncDecl>(DRE->getDecl()))
+            params = FD->getParameterLists().back()->getArray();
+
+        if (params.empty())
+          return { true, E };
+
+        // Extract all arguments.
+        auto *CEA = CE->getArg();
+        if (auto *TSE = dyn_cast<TupleShuffleExpr>(CEA))
+          CEA = TSE->getSubExpr();
+        // The argument is either a ParenExpr or TupleExpr.
+        ArrayRef<Expr *> arguments;
+        ArrayRef<TypeBase *> types;
+
+        if (auto *TE = dyn_cast<TupleExpr>(CEA))
+          arguments = TE->getElements();
+        else if (auto *PE = dyn_cast<ParenExpr>(CEA))
+          arguments = PE->getSubExpr();
+        else
+          arguments = CE->getArg();
+
+        for (auto arg: arguments) {
+          bool isNoEscape = false;
+          while (1) {
+            if (auto AFT = arg->getType()->getAs<AnyFunctionType>()) {
+              isNoEscape = isNoEscape || AFT->isNoEscape();
+            }
+
+            if (auto conv = dyn_cast<ImplicitConversionExpr>(arg))
+              arg = conv->getSubExpr();
+            else if (auto *PE = dyn_cast<ParenExpr>(arg))
+              arg = PE->getSubExpr();
+            else
+              break;
+          }
+          if (!isNoEscape) {
+            unsigned inoutCount = countInOutCaputeres(arg);
+            if (inoutCount) {
+              TC.diagnose(E->getLoc(),
+                diag::nested_function_with_implicit_capture_argument,
+                inoutCount > 1);
+            }
+          }
+        }
+      }
+      return { true, E };
+    }
+  };
+
+  CallWalker walker(TC);
+  for (auto D : llvm::makeArrayRef(SF.Decls).slice(StartElem)) {
+    D->walk(walker);
+  }
+}
```

Comment:
I think we should be able to check this during `applySolution` instead of walking the entire source file again. When we `finishApply` on a call that has a local function `DeclRef` as an argument, we can either check the local function's captures immediately if they've been computed, or else remember whether the function reference was `@noescape` or not and then diagnose when we `computeCaptures` later.